modules/postprocessing/src/processors/depthoffield.cpp

```diff
@@ -0,0 +1,427 @@
+    : Processor()
+    , inport_("inport")
+    , trackingInport_("trackingInport")
+    , outport_("outport")
+    , aperture_("aperture", "Aperture", 0.5f, 0.01f, 1.f, 0.01f)
+    , focusDepth_("focusDepth", "Focus depth", 8.f, 1.0f, 15.0f, 0.1f)
+    , manualFocus_("manualFocus", "Manual focus", true)
+    , approximate_("approximate", "Approximate", false)
+    , viewCountExact_("viewCountExact", "View count", 40, 10, 200)
+    , viewCountApprox_("viewCountApprox", "Rendered view count", 5, 1, 12)
+    , simViewCountApprox_("simViewCountApprox", "Simulated view count", 40, 10, 200)
+    , clickToFocus_(
+          "clickToFocus", "Click to focus", [this](Event* e) { clickToFocus(e); },
+          MouseButton::Left, MouseState::Press, KeyModifier::Control)
+    , camera_("camera", "Camera")
+    , evalCount_(0)
+    , useComputeShaders_(OpenGLCapabilities::getOpenGLVersion() >= 430)
+    , ogCamera_{}
+    , addSampleShader_("fullscreenquad.vert", "dof_exact.frag")
+    , addToLightFieldShader_({{ShaderType::Compute, "dof_approx.comp"}})
+    , averageLightfieldShader_("fullscreenquad.vert", "dof_approx.frag") {
+
+    addPort(inport_);
+    addPort(trackingInport_);
+    addPort(outport_);
+    addProperties(aperture_, focusDepth_, manualFocus_, approximate_, viewCountExact_,
+                  viewCountApprox_, simViewCountApprox_, clickToFocus_, camera_);
+    setApproximate(approximate_);
+
+    trackingInport_.setOptional(true);
+
+    approximate_.onChange([this]() { setApproximate(approximate_); });
+
+    manualFocus_.onChange([this]() {
+        focusDepth_.setReadOnly(!manualFocus_);
+        clickToFocus_.setReadOnly(!manualFocus_);
+    });
+
+    addSampleShader_.onReload([this]() { invalidate(InvalidationLevel::InvalidOutput); });
+    addToLightFieldShader_.onReload([this]() { invalidate(InvalidationLevel::InvalidOutput); });
+    averageLightfieldShader_.onReload([this]() { invalidate(InvalidationLevel::InvalidOutput); });
+}
+
+void DepthOfField::setApproximate(bool approximate) {
+    viewCountExact_.setVisible(!approximate);
+    viewCountApprox_.setVisible(approximate);
+    simViewCountApprox_.setVisible(approximate);
+    if (approximate && !useComputeShaders_) {
+        // Using CPU version of approximative algorithm.
+        LogWarn("Compute shaders are not supported. Approximative depth of field post-processing "
+                << "may be slow.");
+    }
+}
+
+void DepthOfField::clickToFocus(Event* e) {
+    if (!manualFocus_) return;
+
+    auto mouseEvent = static_cast<MouseEvent*>(e);
+    size2_t clickPos(mouseEvent->x(), mouseEvent->y());
+    double depthNdc =
+        inport_.getData()->getDepthLayer()->getRepresentation<LayerRAM>()->getAsNormalizedDouble(
+            clickPos);
+
+    double nearClip = camera_.getNearPlaneDist();
+    double farClip = camera_.getFarPlaneDist();
+    double depthEye = farClip * nearClip / (farClip - depthNdc * (farClip - nearClip));
+    focusDepth_.set(depthEye);
```

Comment:
I see how this would be a good way of abstracting out camera-specific details. However, this change would introduce lots of extra work:
- Converting our vector [0, width-1]x[0, height-1]x[0,1] to [-1,1]x[-1,1]x[-1,1].
- camera::getWorldPosFromNormalizedDeviceCoords itself does more work than we need, including multiple matrix multiplications.
- The result is a world space position, from which we would then need to calculate the distance vector to the camera and project it onto the camera's view vector.

Reply:
Ok. I think that I saw this code pop up a couple of times. Perhaps it should be extracted to a function and reused? I might be mistaken though since there might be small differences

Reply:
Done